dev/spark_overview.md

```diff
@@ -0,0 +1,452 @@
+$ export SPARK_HOME=~/spark-3.3.1-bin-hadoop3
+$ export PATH=$PATH:$SPARK_HOME/sbin
+```
+
+**Warning**: Setting `SPARK_HOME` will affect `pyspark` operation in local mode.
+
+### Customize Spark configuration settings
+```
+$ cp $SPARK_HOME/conf/spark-defaults.conf.template $SPARK_HOME/conf/spark-defaults.conf
+$ cp $SPARK_HOME/conf/spark-env.sh.template $SPARK_HOME/conf/spark-env.sh
+```
+Set `spark.driver.memory` and `spark.driver.maxResultSize` in `spark-defaults.conf` to the maximum
+values that you expect to pull from Spark to Python, such as if you call `df.toPandas()`.
+`1g` is probably reasonable.
+
+Set `spark.sql.shuffle.partitions` to 1-4x the number of cores in your system.
+
+Set `spark.executor.cores` and `spark.executor.memory` to numbers that allow creation of your
+desired number of executors. Those are per-executor values.
+
+### Start the Spark processes
+```
+$SPARK_HOME/sbin/start-master.sh
+```
+Start a worker with this command. Give the worker as much memory as you can afford. You can also
+configure this in `spark-env.sh`.
+```
+$SPARK_HOME/sbin/start-worker.sh -m 16g spark://$(hostname):7077
+```
+
+### Spark UI
+Open http://localhost:8080/ in your browser and copy the cluster URL and port. It will be
+something like `spark://$(hostname):7077`.
+
+If you add the `sbin` to your `PATH` environment variable, here is a one-liner:
+```
+$ start-master.sh && start-worker.sh -m 24g spark://$(hostname):7077
+```
+
+Monitor cluster tasks in your browser.
+
+## Stop the Spark processes
+Stop all of the processes when you complete your work.
+```
+$ stop-worker.sh && stop-master.sh
+```
+
+## Installing a Spark Standalone Cluster on an HPC
+This section describes how you can run Spark jobs on any number of HPC compute nodes.
+
+The HPC GitHub repository contains scripts that will create an ephemeral Spark cluster
+on compute nodes that you allocate.
+
+1. The repository as setup instructions that are not currently correct because the branch
+   is still under review. Follow
+   https://github.com/daniel-thom/HPC/blob/apache-spark-clusters/applications/spark/README.md
+   except substitute step 1 with the steps below.
+
+   Secondly, the README has generic instructions to run Spark in a variety of ways. The rest
+   of this section gives specific guidance on how to apply those instructions to use dsgrid with
+   a Spark cluster in an interactive session.
+```
+$ git clone https://github.com/daniel-thom/HPC.git
+$ cd HPC
+$ git fetch origin apache-spark-clusters
+$ git checkout apache-spark-clusters
+```
+
+2. You probably want to choose compute node(s) with fast local storage. This example will work.
+```
+$ salloc -t 01:00:00 -N1 --account=dsgrid --partition=debug --mem=730G
+```
+Notes:
+- Eagle will let you allocate two debug jobs each with two nodes. So, you can use these scripts
+  to create a four-node cluster for one hour.
+- If the debug partition is not too full, you can append `--qos=standby` to the command above
```

Comment:
If the debug nodes are not in high demand?

Reply:
I've observed that if there are somewhere around 7-9 debug nodes in use (out of 13-14) then I can't get one on standby. The rules aren't documented. Hence, "not too full."